### PR TITLE
Fix union types getting restricted by compiler plugin

### PR DESCRIPTION
**Pull request by dilanSachi**
## Purpose
$subject
Fixes https://github.com/ballerina-platform/ballerina-standard-library/issues/3929
## Checklist
- [x] Linked to an issue
- [x] Updated the changelog
- [x] Added tests
- [ ] Updated the spec
- [x] Checked native-image compatibility

### CODE REVIEW

**Review by chamil321 — APPROVED** [review]

**Review by TharmiganK — COMMENTED** [review]
Can we also add some ballerina tests to make sure the functionality also works?

**Comment by dilanSachi** [comment]
> Can we also add some ballerina tests to make sure the functionality also works?

added a test case

File: ballerina-tests/http-service-tests/tests/anydata_union_return_type_service.bal [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+import ballerina/http;
+import ballerina/test;
+import http_service_tests.records;
```
Comment (TharmiganK) [inline ballerina-tests/http-service-tests/tests/anydata_union_return_type_service.bal:19]: Any specific reason to add the records to a separate module?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java [hunk 2]
```diff
@@ -149,7 +149,7 @@ public static void validateReturnType(SyntaxNodeAnalysisContext ctx, Node node,
         } else if (kind == TypeDescKind.ARRAY) {
             TypeSymbol memberTypeDescriptor = ((ArrayTypeSymbol) returnTypeSymbol).memberTypeDescriptor();
             validateArrayElementTypeInReturnType(
-                    ctx, node, returnTypeStringValue, memberTypeDescriptor, diagnosticCode);
+                    ctx, node, returnTypeStringValue, memberTypeDescriptor, diagnosticCode, isInterceptorType);
```
Comment (TharmiganK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java:152]: Passing `isInterceptorType` along with the `validateArrayElementTypeInReturnType`, will allow developer to pass an array of `http:NextService` or an array of type which has a member `http:NextService`. So, lets remove that.

Seems like we are passing it in other places as well, like when the type is a map (Line #165). Can you change these as well?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java [hunk 3]
```diff
@@ -184,25 +184,31 @@ private static boolean isServiceType(TypeSymbol returnTypeSymbol) {
...
+            validateArrayElementTypeInReturnType(ctx, node, typeStringValue, typeSymbol,
+                    diagnosticCode, isInterceptorType);
         } else if (kind == TypeDescKind.TYPE_REFERENCE) {
             TypeSymbol typeDescriptor = ((TypeReferenceTypeSymbol) memberTypeDescriptor).typeDescriptor();
             TypeDescKind typeDescKind = retrieveEffectiveTypeDesc(typeDescriptor);
-            if (typeDescKind != TypeDescKind.RECORD && typeDescKind != TypeDescKind.MAP &&
-                    typeDescKind != TypeDescKind.TABLE) {
-                reportInvalidReturnType(ctx, node, typeStringValue, diagnosticCode);
+            if (typeDescKind == TypeDescKind.OBJECT) {
+                if (!isHttpModuleType(RESPONSE_OBJ_NAME, typeDescriptor)) {
```
Comment (TharmiganK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java:202]: We don't allow array of `http:Response`. IIRC incase of `Object` type, we only allow `http:Response`. So in this case we can simply call `reportInvalidReturnType` for `Object` types.
@chamil321 can you also verify?

File: ballerina-tests/http-service-tests/tests/anydata_union_return_type_service.bal (hunk 1 above)
Comment (dilanSachi) [reply ballerina-tests/http-service-tests/tests/anydata_union_return_type_service.bal:19]: Added that to simulate the original use case

Comment (TharmiganK) [reply ballerina-tests/http-service-tests/tests/anydata_union_return_type_service.bal:19]: Ah ok, in that case can we also run the `GraalVM Check` on your branch to see whether this is working without any issue?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java (hunk 3 above)
Comment (chamil321) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java:202]: Yes, we only return the `http:Response` as an object from the resource.

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java (hunk 2 above)
Comment (dilanSachi) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java]: Updated

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java (hunk 3 above)
Comment (dilanSachi) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java]: Updated

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java (hunk 2 above)
Comment (dilanSachi) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java]: Passing https://github.com/dilanSachi/module-ballerina-http/actions/runs/4013386586

File: ballerina-tests/http-service-tests/tests/anydata_union_return_type_service.bal (hunk 1 above)
Comment (dilanSachi) [reply ballerina-tests/http-service-tests/tests/anydata_union_return_type_service.bal:19]: Passing https://github.com/dilanSachi/module-ballerina-http/actions/runs/4013386586

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java [hunk 4]
```diff
@@ -192,22 +192,58 @@ private static void validateArrayElementTypeInReturnType(SyntaxNodeAnalysisConte
...
+                    if (effectiveTypeDesc == TypeDescKind.OBJECT || effectiveTypeDesc == TypeDescKind.ERROR) {
+                        reportInvalidReturnType(ctx, node, typeStringValue, diagnosticCode);
+                        return;
+                    } else if (effectiveTypeDesc == TypeDescKind.TYPE_REFERENCE) {
+                        if (isStatusCodeResponseType((TypeReferenceTypeSymbol) typeSymbol)) {
+                            reportInvalidReturnType(ctx, node, typeStringValue, diagnosticCode);
+                            return;
+                        }
+                        Optional<String> definitionName = ((TypeReferenceTypeSymbol) typeSymbol).definition().getName();
+                        if (definitionName.isPresent() && definitionName.get().equals("StatusCodeResponse")) {
```
Comment (TharmiganK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java:219]: What is the purpose of this check? Does this checks the `StatusCodeResponse` included records? i.e.
```bal
type AlbumConflict record {|
    *http:Conflict;
    record {
        string message;
    } body;
|};
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java [hunk 5]
```diff
@@ -192,22 +192,58 @@ private static void validateArrayElementTypeInReturnType(SyntaxNodeAnalysisConte
...
+            validateArrayElementTypeInReturnType(ctx, node, typeStringValue, memberTypeDescriptor,
+                    diagnosticCode);
         } else {
             reportInvalidReturnType(ctx, node, typeStringValue, diagnosticCode);
         }
     }
 
+    public static boolean isStatusCodeResponseType(TypeReferenceTypeSymbol typeSymbol) {
+        if (typeSymbol.getName().isPresent()) {
+            return typeSymbol.getName().get().equals("StatusCodeResponse");
```
Comment (TharmiganK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java:242]: Can we make this `StatusCodeResponse` string as a constant and reuse?

Comment (TharmiganK) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java:242]: Also we should check whether this `StatusCodeResponse` record is from `http` package. So can't we reuse the `isHttpModuleType` function here?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java (hunk 4 above)
Comment (dilanSachi) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java:219]: It only checks for `StatusCodeResponse` type

File: compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java (hunk 5 above)
Comment (dilanSachi) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java]: Updated with the suggestion

Comment (dilanSachi) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/http/compiler/HttpCompilerPluginUtil.java]: Updated with the suggestion

**Review by TharmiganK — APPROVED** [review]

**Pull request merged**